internal/browser/browser.go

```diff
@@ -0,0 +1,376 @@
+	if t.lastIteration != nil {
+		userOnLastPage := newYOffset > t.lastIteration.totalProcessedLines-t.termdbmsTable.Viewport.Height
+		if newYOffset > oldYOffset {
+			if userOnLastPage {
+				cmd = tea.Batch(cmd, func() tea.Msg {
+					return FetchMoreRowsMsg{}
+				})
+			}
+		}
+
+	}
+	return t, cmd
+}
+
+type SqlIterator struct {
+	rows                *sql.Rows
+	resultPipe          chan []interface{}
+	rowsFinished        bool
+	totalProcessedLines int
+	columnNames         []string // cannot access these after rows.Close(), hence save them
+	Iterating           bool
+}
+
+func NewSqlIterator(maxIterationCount int, rows *sql.Rows) *SqlIterator {
+	var si SqlIterator
+	si.rows = rows
+	si.columnNames, _ = rows.Columns()
+	si.resultPipe = make(chan []interface{}, maxIterationCount+1)
+	go si.Iterate(maxIterationCount)
+	return &si
+}
+
+func (si *SqlIterator) Iterate(maxIterationCount int) {
+	si.Iterating = true
+	var i int
+	for i = 0; si.rows.Next() && i < maxIterationCount; i++ { // each row of the table
+		// golang wizardry
+		columns := make([]interface{}, len(si.columnNames))
+		columnPointers := make([]interface{}, len(si.columnNames))
+		// init interface array
+		for i := range columns {
+			columnPointers[i] = &columns[i]
+		}
+		if err := si.rows.Scan(columnPointers...); err != nil {
+			break
+		}
+		si.resultPipe <- columnPointers
+	}
+	if i != maxIterationCount { // means query finished and there will be no more results
+		close(si.resultPipe)
+	}
+	si.Iterating = false
+}
+
+type NewRowsMessage [][]interface{}
+
+func (si *SqlIterator) IterateCmd(timeline time.Duration) func() tea.Msg {
+	return func() tea.Msg {
+		var newRows [][]interface{}
+		var timeout bool
+		for {
+			select {
+			case row := <-si.resultPipe:
+				if row == nil {
+					si.rowsFinished = true
+					return NewRowsMessage(newRows)
+				}
+				si.totalProcessedLines++
+				newRows = append(newRows, row)
+			case <-time.After(timeline):
+				timeout = true
+				break
+			}
+			if timeout {
+				break
+			}
+		}
+		return NewRowsMessage(newRows)
+	}
+}
+
+func (m *table) PopulateDataForResult(rows [][]interface{}) {
+	columnNames := m.lastIteration.columnNames
+	columnValues := make(map[string][]interface{})
+	if m.termdbmsTable.QueryResult != nil && m.termdbmsTable.QueryData != nil {
+		if m.termdbmsTable.QueryResult.Data["0"] != nil {
+			columnValues = m.termdbmsTable.QueryResult.Data["0"].(map[string][]interface{})
+		}
+	}
+
+	for _, row := range rows {
+		for i, colName := range columnNames {
+			val := row[i].(*interface{})
+			columnValues[colName] = append(columnValues[colName], *val)
+		}
+	}
+
+	// onto the next schema
+	if m.termdbmsTable.QueryResult != nil && m.termdbmsTable.QueryData != nil {
+		m.termdbmsTable.QueryResult.Data["0"] = columnValues
+		m.termdbmsTable.QueryData.TableHeaders["0"] = columnNames // headers for the schema, for later reference
+		m.termdbmsTable.QueryData.TableIndexMap[1] = "0"
```

Comment:
There are lots of unneeded structures from termdbms that we do not utilize. I agree it would be great to refactor them. Can we address them in another PR if that works for you?

Reply:
OK.